lib/core/models/repositories/roleRepository.js

```diff
@@ -181,25 +195,80 @@ class RoleRepository extends Repository {
    * @returns Promise
    */
   validateAndSaveRole (role, options) {
-    return role.validateDefinition()
+    return role
+      .validateDefinition()
       .then(() => {
         if (role._id === 'anonymous' && !role.canLogIn()) {
           errorsManager.throw('security', 'role', 'login_required');
         }
 
-        this.kuzzle.emit(
-          'core:roleRepository:save',
-          { _id: role._id, controllers: role.controllers });
-
-        return this.persistToDatabase(role, options);
+        this.kuzzle.emit('core:roleRepository:save', {
+          _id: role._id,
+          controllers: role.controllers
+        });
       })
+      .then(() => this.checkRoleControllersAndActions(role))
+      .then(() => this.persistToDatabase(role, options))
       .then(() => this.loadOneFromDatabase(role._id))
       .then(updatedRole => {
         this.roles.set(role._id, updatedRole);
         return updatedRole;
       });
   }
 
+  /**
+   * Given a Role object, checks if its controllers and actions exist.
+   *
+   * @param {Role} role
+   * @returns Promise
+   */
+  checkRoleControllersAndActions(role) {
```

Comment:
Nit/proposal: could be renamed because it should checks only for native action existence